runtime/lua/vim/lsp/protocol.lua

```diff
@@ -923,6 +923,28 @@ function protocol.resolve_capabilities(server_capabilities)
     error("The server sent invalid codeActionProvider")
   end
 
+  if server_capabilities.declarationProvider == nil then
+    general_properties.declaration = false
+  elseif type(server_capabilities.declarationProvider) == 'boolean' then
+    general_properties.declaration = server_capabilities.declarationProvider
+  elseif type(server_capabilities.declarationProvider) == 'table' then
+    -- TODO: support more detailed declarationProvider options.
+    general_properties.declaration = false
```

Comment:
Hmm, when we do this, doesn't it disable these requests from happening? Should we default to true and see what happens instead?

Reply:
My understanding is that the block that deals with when `type(server_capabilities.declarationProvider) == table` hasn't been fully implemented yet. So maybe in a separate PR that can be redone properly.

For the other branches, I think neovim should not "default to `true` and see what happens". As an example, LanguageServer.jl specifies it does not support some features and throws an exception and crashes if a request is made that is not supported: https://github.com/julia-vscode/LanguageServer.jl/issues/736. I'm sure other language server have similar undefined behavior when this occurs.

Reply:
See the following for when `type(server_capabilities.implementationProvider) == table`, where it defaults to `false`.

https://github.com/kdheepak/neovim/blob/c37281f32fd49d1cada353d0b7ad95e42c1af0f2/runtime/lua/vim/lsp/protocol.lua#L954

Reply:
I think some of those are not exactly correct, but we could probably do as you suggested.

I am looking at the registration options and how they work, some of the things we do support (for example, linkSupport and others). This shouldn't be up to you to solve though probably :)

Reply:
When it returns a table, it means it does support the protocol, is my understanding. However, it gives additional information about certain registration options, incremental updates, etc.

Reply:
Yes, I agree, `table` does mean it supports the protocol and the options need to be parsed correctly.

Reply:
Are you interested in parsing the table or not for these requests? Just trying to gauge your interest :)

Reply:
I'd be interested in contributing! But as a separate PR? This currently crashes the Julia Language Server, and it'll be easier to convince colleagues to try to nightly nvim if that didn't happen :)

Reply:
:+1: Sounds good. Let me review again and I'll merge I think.